opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingHelper.java

```diff
@@ -0,0 +1,285 @@
+
+    private static Map<Object,Integer> state = Collections.synchronizedMap(new WeakHashMap<Object,Integer>());
+
+    public OpenTracingHelper(Rule rule) {
+        super(rule);
+    }
+
+    /**
+     * This method returns the OpenTracing tracer.
+     *
+     * @return The tracer
+     */
+    public Tracer getTracer() {
+        return tracer;
+    }
+
+    /**
+     * This method requests that the supplied span becomes the
+     * current 'active' span.
+     *
+     * @param span The span
+     */
+    public void activateSpan(Span span) {
+        spanManager.activate(span);
+    }
+
+    /**
+     * This method returns the currently active span.
+     *
+     * @return The current span, or null if no active span exists
+     */
+    public Span currentSpan() {
+        return spanManager.current().getSpan();
+    }
+
+    /**
+     * This method requests that the current active span
+     * should be deactivated. If the underlying
+     * span management mechanism maintains a stack, then
+     * this will result in the previous (unfinished) parent
+     * span being reinstated as the current active span.
+     *
+     * @return The span being deactivated, or null if no active span found
+     */
+    public Span deactivateCurrentSpan() {
+        ManagedSpan current = spanManager.current();
+        current.deactivate();
+        return current.getSpan();
+    }
+
+    /**
+     * This method establishes an association between an application object
+     * (i.e. the subject of the instrumentation) and a span. Once the
+     * application object is no longer being used, the association with the
+     * span will automatically be discarded.
+     *
+     * @param obj The application object to be associated with the span
+     * @param span The span
+     */
+    public void associateSpan(Object obj, Span span) {
+        spanAssociations.put(obj, span);
+    }
+
+    /**
+     * This method retrieves the span associated with the supplied application
+     * object.
+     *
+     * @param obj The application object
+     * @return The span, or null if no associated span exists
+     */
+    public Span retrieveSpan(Object obj) {
+        return spanAssociations.get(obj);
+    }
+
+    /**********************************************/
+    /** Needs to be replaced by span.isFinished() */
+    public void finishedSpan(Object key, Span span) {
+        finished.put(key, span);
+    }
+
+    public boolean isFinished(Object key) {
+        return finished.containsKey(key);
+    }
+    /**********************************************/
+
+    /**
+     * This method enables an instrumentation rule to record a 'state'
+     * number against an application object. This can be used in
+     * subsequent rules to determine what the valid actions that
+     * can be performed.
+     *
+     * @param obj The application object
+     * @param value The state value
+     */
+    public void setState(Object obj, int value) {
```

Comment:
It is used in the `jetty.btm` rules as the instrumentation point used by the rules is called multiple times, but we only want to install the `TracingFilter` once. So in this particular case it is a binary decision, so could have an 'isInstalled' mechanism, but thought it may be more general purpose if it was possible to represent different states, in case there are situations where more than two states are required.

I'll add more detail to the javadoc.